config/7.1.0/securebanking/ig/scripts/groovy/CheckIdempotencyKey.groovy

```diff
@@ -0,0 +1,70 @@
+/*
+ * Script to check the idempotency
+ */
+
+import groovy.json.JsonOutput
+
+def fapiInteractionId = request.getHeaders().getFirst("x-fapi-interaction-id");
+if (fapiInteractionId == null) fapiInteractionId = "No x-fapi-interaction-id";
+SCRIPT_NAME = "[CheckIdempotencyKey] (" + fapiInteractionId + ") - ";
+
+logger.debug(SCRIPT_NAME + "Running...")
+
+// response object
+responseCheck = new Response(Status.OK)
+responseCheck.headers['Content-Type'] = "application/json"
+
+String idempotencyKeyHeaderValue = request.getHeaders().getFirst("x-idempotency-key")
+logger.debug(SCRIPT_NAME + "IdempotencyKey header value: " + idempotencyKeyHeaderValue)
+
+def method = request.method
+
+def filter = "_queryFilter="+ idempotencyArgFieldToFilter + "+eq+%22" + URLEncoder.encode(idempotencyKeyHeaderValue, "UTF-8") + "%22"
```

Comment:
The key only needs to be unique to the TPP, not globally unique. Otherwise values sent by different TPPs could clash, we can't assume how they will implement the idempotency key generation.

Reply:
If you mean the argument `idempotencyArgFieldToFilter` is not the key the key sent by the TPP is `URLEncoder.encode(idempotencyKeyHeaderValue, "UTF-8")` obtained from the request few lines above, `idempotencyArgFieldToFilter` is the field name where we store the key send by the TPP, for example for any payment initiation we store the key in the field `IdempotencyKey`, the file payment initiation is a little bit different because the submission of the file content need to be idempotent too, for that reason in that case we have two different fields, one for the consent named `IdempotencyKey` (the same as other initiations) and the specific field named `IdempotencyKeyFile` to store the idempotency key send by the TPP when the file content is submitted.

Reply:
I may be misunderstanding the logic somewhere but here is an example scenario which I think will behave incorrectly with the current impl.

- dan-tpp apiClient sends a DomesticPaymentConsent with idempotencyKey = "1"
- jordi-tpp apiClient sends a DomesticPaymentConsent with idempotencyKey = "1"

What response does jordi-tpp get? From my reading of the code then you will get dan-tpp's consent sent back to you.

Each TPP's idempotencyKey must be treated as an independent set of values.

Reply:
Got It! we need add to the filter the apiClient too to request only the consents linked with the TPP to avoid retrieve the consent of another TPP filtering only by idempotency-key.

Reply:
Done!